### ISSUE TITLE
Add support for format_parameters to the new fastAPI endpoints

### ISSUE THREAD

**Issue by amoghrajesh** [apache/airflow#43872]
### Description

Legacy API has support for format_parameters which is used to place restrictions on the response page. For example: `@format_parameters({"limit": check_limit})` we can set limit on the limit of entries returned by an API.

### Use case/motivation

Cleaner migration from legacy to fastapi. Will be useful for all endpoints.

### Related issues

Child of https://github.com/apache/airflow/issues/42370

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by rawwar** [comment]
One question here: Wouldn't this confuse users when they specify "limit" as part of the request body/param, unaware of the maximum_page_limit? As in, if limit > maximum_page_limit, the API response returns the `maximum_page_limit` number of records. The user needs to refer to the warning logs to understand that the reason for the truncation is that they have set `maximum_page_limit`.   I think, along with the warning log, we should also update REST API documentation.

**Comment by kandharvishnu** [comment]
I am interested in working on this issue

**Comment by amoghrajesh** [comment]
> Wouldn't this confuse users when they specify "limit" as part of the request body/param, unaware of the maximum_page_limit? As in, if limit > maximum_page_limit, the API response returns the maximum_page_limit number of records. The user needs to refer to the warning logs to understand that the reason for the truncation is that they have set maximum_page_limit. I think, along with the warning log, we should also update REST API documentation.

Yeah, that's a fair point. However, we do mention this in the API doc https://github.com/apache/airflow/blob/f7da5e48a2740a2bfb1c3dd57bb463ae29599a26/docs/apache-airflow/security/api.rst#page-size-limit about default size. I am all for adding it for clarification purpose, but probably that should be a different issue, unrelated to this @rawwar?

**Comment by potiuk** [comment]
> Agree. The limit was implemented in a bit to "silent" way in the old  API and it created confusion. 

Even if we have the "max limit" internally - I think best thing to respond with error when:

a) you do not specify limit and number of responses exceeds the "max limit" 
b) you specify limit that is bigger than "max-limit"

Both cases should result in ERROR - currently both will silently return you something different than you expected.

**Comment by rawwar** [comment]
> Both cases should result in ERROR - currently, both will silently return you something different than you expected.

+1 to returning an Error. FYI, @pierrejeambrun

**Comment by pierrejeambrun** [comment]
I'm not sure I understand correctly.

Is this issue for adding a generic `@format_parameters` decorator, or something to specifically the `limit` case ? 

If it's the former, I don't see what advantages it brings in regards to Pydantic + FastAPI query param validation schema ? All of the parameter validation is already decoupled from the endpoint logic in it's current form.

**Comment by rawwar** [comment]
@pierrejeambrun , I think @amoghrajesh created it for a generic `@format_parameters`. I started discussing about `maximum_page_limit` as a specific case.

**Comment by pierrejeambrun** [comment]
Maybe I am missing something. Can you elaborate the idea and highlight what advantages it would brings to develop this custom decorator for FastAPI compared to the validation offered natively by pydantic `Annotated` + `BeforeValidator/PostValidator` + `field/model` validators that is directly integrated with FastAPI ?

**Comment by pierrejeambrun** [comment]
Closing for now. With FastAPI dependency system + Pydantic validation, we can basically achieve that just by python type annotation, we don't even need a decorator.
